### ISSUE TITLE
Behavior of CMAKE_BUILD_TYPE=Release is weird - doesn't include standard libraries such as nvim.lua

### ISSUE THREAD

**Issue by jdix531** [neovim/neovim#21190]
### Neovim version (nvim -v)

stable (v0.8.1)

### Vim (not Nvim) behaves the same?

n/a

### Operating system/version

ubuntu (kde aka kubuntu)

### Terminal name/version

konsole bash 5.1.16

### $TERM environment variable

xterm-256color

### Installation

build from source

### How to reproduce the issue

I built from source using:

`make CMAKE_BUILD_TYPE=Release CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=$HOME/.local"`

which led to a functional installation until I tried to use certain core libraries such as nvim.lua at which point there were not able to be found. Specifically, I was trying to use AstroNvim config and get errors running PackerSync.

### Expected behavior

No errors, the core libs should always be included (I would think, or if there's a switch I don't see it in docs or my debug build full-version output).

When running: `make CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=$HOME/.local"` I don't get errors when trying to use configs that rely on the core libs.

### Actual behavior

I get the below error output that leads me to think it is nvim error not AstroNvim error:

```Error detected while processing /home/josh/.config/nvim/init.lua:
Failed to load core.utils
vim/_init_packages.lua:0: module 'vim.lsp' not found:
^Ino field package.preload['vim.lsp']
^Ino file './vim/lsp.lua'
^Ino file '/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/share/luajit-2.1.0-beta3/vim/lsp.lua'
^Ino file '/usr/local/share/lua/5.1/vim/lsp.lua'
^Ino file '/usr/local/share/lua/5.1/vim/lsp/init.lua'
^Ino file '/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/share/lua/5.1/vim/lsp.lua'
^Ino file '/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/share/lua/5.1/vim/lsp/init.lua'
^Ino file './vim/lsp.so'
^Ino file '/usr/local/lib/lua/5.1/vim/lsp.so'
^Ino file '/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/lib/lua/5.1/vim/lsp.so'
^Ino file '/usr/local/lib/lua/5.1/loadall.so'
^Ino file './vim.so'
^Ino file '/usr/local/lib/lua/5.1/vim.so'
^Ino file '/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/lib/lua/5.1/vim.so'
^Ino file '/usr/local/lib/lua/5.1/loadall.so'
Failed to load core.diagnostics

vim/_init_packages.lua:0: module 'vim.diagnostic' not found:
^Ino field package.preload['vim.diagnostic']
^Ino file './vim/diagnostic.lua'
^Ino file '/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/share/luajit-2.1.0-beta3/vim/diagnostic.lua'
^Ino file '/usr/local/share/lua/5.1/vim/diagnostic.lua'
^Ino file '/usr/local/share/lua/5.1/vim/diagnostic/init.lua'
^Ino file '/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/share/lua/5.1/vim/diagnostic.lua'
^Ino file '/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/share/lua/5.1/vim/diagnostic/init.lua'
^Ino file '/home/josh/.cache/nvim/packer_hererocks/2.1.0-beta3/share/lua/5.1/vim/diagnostic.lua'
^Ino file '/home/josh/.cache/nvim/packer_hererocks/2.1.0-beta3/share/lua/5.1/vim/diagnostic/init.lua'
^Ino file '/home/josh/.cache/nvim/packer_hererocks/2.1.0-beta3/lib/luarocks/rocks-5.1/vim/diagnostic.lua'
^Ino file '/home/josh/.cache/nvim/packer_hererocks/2.1.0-beta3/lib/luarocks/rocks-5.1/vim/diagnostic/init.lua'
^Ino file './vim/diagnostic.so'
^Ino file '/usr/local/lib/lua/5.1/vim/diagnostic.so'
^Ino file '/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/lib/lua/5.1/vim/diagnostic.so'
^Ino file '/usr/local/lib/lua/5.1/loadall.so'
^Ino file '/home/josh/.cache/nvim/packer_hererocks/2.1.0-beta3/lib/lua/5.1/vim/diagnostic.so'
^Ino file './vim.so'
^Ino file '/usr/local/lib/lua/5.1/vim.so'
^Ino file '/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/lib/lua/5.1/vim.so'
^Ino file '/usr/local/lib/lua/5.1/loadall.so'
^Ino file '/home/josh/.cache/nvim/packer_hererocks/2.1.0-beta3/lib/lua/5.1/vim.so'
E484: Can't open file /home/josh/.local/share/nvim/syntax/syntax.vim
```

NOTE: the directory "/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim" is where neovim is cloned to as part of the install process. The full install script I run is below, where I switch the comment line for testing right now:

```
#!/bin/bash

git clone https://github.com/neovim/neovim
cd neovim
git fetch && git pull
git checkout stable
make CMAKE_BUILD_TYPE=Release CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=$HOME/.local"
#make CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=$HOME/.local"
make install
```

NOTE: I am clearing the following directories between nvim installs:

```
rm -rf ~/.local/share/nvim
rm -rf ~/.cache/nvim
rm -rf ~/.local/site/nvim
```

NOTE: The error doesn't appear if I install debug THEN release which leads me to think this is for sure missing modules from the nvim release install side - the only question is:

"Am I missing an install switch to include the core libs, is this by design (and perhaps AstroNvim is making a weird assumption), or is this an issue in the make script?"

Thanks for all the efforts on this great project, let me know if there is any more information needed.

**Comment by clason** [comment]
Did you `make install`?

**Comment by dundargoc** [comment]
Follow up question: did you `make distclean` beforehand?

**Comment by jdix531** [comment]
@clason yes, I included my full script of steps towards the bottom there.
@dundargoc I was removing my cloned neovim folder every time which should have done the same thing according to the docs.

THAT SAID - I stumbled across the 'deps' step - which to be fair seems missing from https://github.com/neovim/neovim/wiki/Installing-Neovim#install-from-source.

So I'm thinking the default `make CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=$HOME/.local"` (in debug mode) installs deps for you but release mode `make CMAKE_BUILD_TYPE=Release CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=$HOME/.local"` doesn't include them.

My issue is then installing deps, I tried:

```
mkdir .deps
cd .deps
cmake ../cmake.deps/
gmake
cd ..
```

But it fails:

```
...
[ 87%] Building C object deps/libuv/CMakeFiles/uv_a.dir/src/unix/random-getrandom.c.o
[ 90%] Building C object deps/libuv/CMakeFiles/uv_a.dir/src/unix/random-sysctl-linux.c.o
[ 92%] Building C object deps/libuv/CMakeFiles/uv_a.dir/src/unix/epoll.c.o
[ 95%] Linking C static library libuv_a.a
[ 95%] Built target uv_a
[ 97%] Building C object CMakeFiles/luv.dir/src/luv.c.o
[100%] Linking C shared module luv.so
[100%] Built target luv
Consolidate compiler generated dependencies of target uv_a
[ 95%] Built target uv_a
Consolidate compiler generated dependencies of target luv
[100%] Built target luv
Install the project...
-- Install configuration: ""
-- Installing: /home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/lib/luarocks/rocks-5.1/luv/1.44.2-1/lib/luv.so
luv 1.44.2-1 is now installed in /home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr (license: Apache 2.0)

nvim-client 0.2.4-1 depends on coxpcall (not installed)
Installing https://luarocks.org/coxpcall-1.17.0-1.src.rock

coxpcall 1.17.0-1 is now installed in /home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr (license: MIT/X11)

/usr/bin/cc -O2 -fPIC -I/home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr/include/luajit-2.1 -c nvim/native.c -o nvim/native.o
/usr/bin/cc -shared -o nvim/native.so nvim/native.o
nvim-client 0.2.4-1 is now installed in /home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim/.deps/usr (license: Apache)

[100%] Built target nvim-client
[100%] Built target clean-shared-libraries
[100%] Built target third-party
cd .deps && \
        cmake -G 'Ninja'   \
         /home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim//cmake.deps
CMake Error: Error: generator : Ninja
Does not match the generator used previously: Unix Makefiles
Either remove the CMakeCache.txt file and CMakeFiles directory or choose a different binary directory.
make: *** [Makefile:108: build/.ran-deps-cmake] Error 1
cd .deps && \
        cmake -G 'Ninja'   \
         /home/josh/code/bootstrap-autoclone/roles/install_nvim/files/neovim//cmake.deps
CMake Error: Error: generator : Ninja
Does not match the generator used previously: Unix Makefiles
Either remove the CMakeCache.txt file and CMakeFiles directory or choose a different binary directory.
make: *** [Makefile:108: build/.ran-deps-cmake] Error 1
```

The docs seem to say that the deps should be downloaded automatically be default (USE_BUNDLED=1) from here: https://github.com/neovim/neovim/wiki/Building-Neovim#building.

So I'm still thinking there's an error with the default instructions for release mode where it should be installing deps but is not.

**Comment by clason** [comment]
No, `deps` are a red herring here. I cannot reproduce your error, either, with the exact steps you gave.

**Comment by jdix531** [comment]
@clason 
Let me try to give better steps:

```
# cleanup:
rm -rf ~/.local/site/nvim
rm -rf ~/.local/share/nvim
rm -rf ~/.cache/nvim
mv ~/.config/nvim ~/.config/nvim.bak
## directory where neovim source is cloned:
rm -rf ~/code/bootstrap-autoclone/roles/install_nvim/files/neovim

# install nvim release mode
git clone https://github.com/neovim/neovim
cd neovim
git fetch && git pull
git checkout stable
make CMAKE_BUILD_TYPE=Release CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=$HOME/.local"
make install

# use AstroNvim config which seems to call these deps and produces the above errors (again, only when installed in release mode for me)
git clone https://github.com/AstroNvim/AstroNvim ~/.config/nvim
nvim
```

And I then get the mentioned errors. You're clearing the directories mentioned right @clason? I mentioned I don't get the issue if I don't clear those directories which I believe is due to previous installs of the deps from the debug version. If you still can't reproduce I will try another machine (this is kubuntu but I have a manjaro install as well).

**Comment by clason** [comment]
Sorry, I can only answer questions about building and installing Neovim. For AstroNvim, you're on your own.

**Comment by jdix531** [comment]
@clason will check that, ty! I'll also see if I can rephrase this as an NVIM problem. I understand not wanting to chase this down.

I don't know if running the make deps steps manually did anything to my system outside the build dir but I have not been able to reproduce the issue since:

```
mkdir .deps
cd .deps
cmake ../cmake.deps/
gmake
cd ..
```

So this looks good to close anyways. Weird I switched back and forth between bad/good state several times now can't reproduce - not sure if that code there is related.
